Updated cURL library to support static and dynamic version using the …

jvelilla opened the pull request:
…same API:

By default the mode is dynamic, to use static
set the variable static_curl_enabled, with value true in the client library.

jvelilla committed:
Updated cURL library to support static and dynamic version using the same API:
By default the mode is dynamic, to use static
set the variable static_curl_enabled, with value true in the client library.

jvelilla committed:
Updated cURL static library with MT curl static option.
Updated CURL_FUNCTION with is_static precondition.

jvelilla committed:
Updated api_wrapper library with a MODULE_LOADER abstraction to support
static and dynamic mode.
Updated cURL code, clean code.

jvelilla committed:
Updated cURL simplified code

jocelyn commented on Src/library/api_wrapper/api/module_loader.e:
```diff
@@ -0,0 +1,36 @@
+note
+	description: "Generic {MODULE_LOADER} representing an dynamic loaded library or static librart."
```
typo  
- "an dynamic..." -> "a dynamic..."
- "librart"

jvelilla replied:
fixed

jocelyn commented on Src/library/api_wrapper/api/static_module.e:
```diff
@@ -0,0 +1,46 @@
+feature -- Query
+
+	api_pointer (a_api_name: READABLE_STRING_8): POINTER
+			-- Retrieve an API function/variable pointer given an API name.
```
mentions `a_api_name` in comment.

jvelilla replied:
fixed

jocelyn commented on Src/library/api_wrapper/api/static_module.e:
```diff
@@ -0,0 +1,46 @@
+note
+	description: "Objec representing {STATIC_MODULE}."
```
typo   `Objec` ... also I am sure we could have a more useful comment.

jvelilla replied:
fixed

jocelyn commented on Src/library/cURL/cURL.ecf:
```diff
@@ -23,38 +23,60 @@
 		</external_object>
 		<external_object location="$(ISE_LIBRARY)\library\cURL\spec\$(ISE_C_COMPILER)\$(ISE_PLATFORM)\lib\ileiffel_curl.lib">
 			<condition>
+				<custom name="static_curl_enabled" excluded_value="true"/>
```
I would suggest to put in the .ecf description (that we usually don't use enough) explanation about `static_curl_enabled` and also precise that by default `dynamic` behavior is used.

jvelilla replied:
Something like

Set `static_curl_enabled` to `true` in the `.ecf` of your app if you want to use cURL in static mode, by default `dynamic` mode is used.

Example:
```
<variable name="static_curl_enabled" value="true"/>
```

jocelyn commented on Src/library/cURL/curl_easy_externals.e:
```diff
@@ -17,14 +17,12 @@ feature -- Command
-			if l_api /= default_pointer then
-				Result := c_init (l_api)
-			end
+			Result := c_init (l_api)
```
What if l_api is default pointer and dynamic lib is the default ?

jocelyn replied:
the code violates the precondition of c_init ..

same comment for all code interfacing with the following cURL C externals

jvelilla replied:
you are right, that's why in the previous commit  I've done something like this

```
l_api := api_loader.api_pointer ("curl_easy_init")
if l_api /= default_pointer then
    Result := c_init (l_api)
else
    Result := c_init (default_pointer)
end
```	

But, we can check precondition so we will be safe, so we can write something like


```
l_api := api_loader.api_pointer ("curl_easy_init")
if l_api /= default_pointer  or is_static then
    Result := c_init (l_api)
end
```	
What do you think?

jocelyn reviewed (commented):
I started to review this approach.
But like that I am not a big fan, as STATIC_MODULE sounds like a hack.

We may think that api_pointer would return the expected function pointer .

I will review the other approach now, but for now I post those first comments.

jvelilla commented:
Yes, STATIC_MODULE it is a hack, that enables us to use the same pattern of code when we have static or dynamic C library.
For static mode there is no API pointer, that's why we return a default_pointer.

jvelilla committed:
Updated cURL library
Fixed typos and improve comments.
Updated ecf file, udpated description describing the dynamic and static modes
Added linux option to use cURL with static mode.
Updated code, checking the preconditions before to call the c externals features.

manuseiffel commented on Src/library/cURL/curl_easy_externals.e:
```diff
@@ -17,12 +17,12 @@ feature -- Command
 		do
 			l_api := api_loader.api_pointer ("curl_easy_init")
-			if l_api /= default_pointer then
+			if l_api /= default_pointer or is_static then
```
I don't think we need the `if` statement anymore.

jvelilla replied:
I just keep the `if ` statement to check the precondition for the external feature.

manuseiffel replied:
No need to check the precondition since it is already there. You could write `check l_api /= default_pointer or is_static end` but no need for the if.

jvelilla committed:
Updated code using check instructions like this
`check l_api /= default_pointer or is_static end`
since there is no need to check the preconditions since it
is already there.
